src/GPS.cpp

```diff
@@ -0,0 +1,120 @@
+#include "GPS.hpp"
+#include <nmea0183/GPS.hpp>
+
+using namespace nmea0183;
+using namespace marnav;
+using namespace std;
+using namespace gps_base;
+
+GPS_SOLUTION_TYPES GPS::getPositionType(utils::optional<nmea::mode_indicator> mode)
+{
+    switch (*mode) {
+        case nmea::mode_indicator::invalid:
+            return GPS_SOLUTION_TYPES::INVALID;
+        case nmea::mode_indicator::autonomous:
+            return GPS_SOLUTION_TYPES::AUTONOMOUS;
+        case nmea::mode_indicator::differential:
+            return GPS_SOLUTION_TYPES::DIFFERENTIAL;
+        case nmea::mode_indicator::estimated:
+            return GPS_SOLUTION_TYPES::INVALID;
+        case nmea::mode_indicator::manual_input:
+            return GPS_SOLUTION_TYPES::INVALID;
+        case nmea::mode_indicator::simulated:
+            return GPS_SOLUTION_TYPES::INVALID;
+        case nmea::mode_indicator::data_not_valid:
+            return GPS_SOLUTION_TYPES::INVALID;
+        case nmea::mode_indicator::precise:
+            return GPS_SOLUTION_TYPES::AUTONOMOUS;
+        default:
+            return GPS_SOLUTION_TYPES::INVALID;
+    }
+}
+
+base::Time GPS::buildRockTime(utils::optional<nmea::time> const& optional_time,
+    utils::optional<nmea::date> const& optional_date)
+{
+    if (optional_date.has_value() && optional_time.has_value()) {
+        auto date = optional_date.value();
+        auto time = optional_time.value();
+        return base::Time::fromTimeValues(date.year() + 2000,
+            static_cast<int>(date.mon()),
+            date.day(),
+            time.hour(),
+            time.minutes(),
+            time.seconds(),
+            time.milliseconds(),
+            0);
+    }
+    else {
+        return base::Time::now();
```

Comment:
This is a point we could discuss as I am not sure whats the best solution. Is it better to return `now` or base::Time()? I think it boils down to how would we expect to use the time from this data, so I am going towards using `now` here.

Reply:
Thinking a bit about it, I think we could always use Time::now, no? For log analysis, using the message time could mean that this data is not synchronized with the rest of the logs, if there is a synchronization point in the device, we could avoid this kind of issue. Then, conditioning this to the presence of this kind of data in the message, and returning Time::now when it is not present would necessarily lean on the synchronization between Time::now and the GPS time, which just makes the whole point of reading the GPS time moot. Do you agree?

Reply:
yes, this is how I did at first